#include "mainwindow.h"
#include "ui_mainwindow.h"

#include <QStandardItemModel>
#include <QHeaderView>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent)
    , ui(new Ui::MainWindow)
{
    ui->setupUi(this);

    // ØªØ­Ù…ÙŠÙ„ Ø£ÙŠ Ø¬Ø¯Ø§ÙˆÙ„ Ù…ÙˆØ¬ÙˆØ¯Ø© Ø¹Ù„Ù‰ Ø§Ù„Ù‡Ø§Ø±Ø¯
    db.loadAllTables("data");

    // Ø±Ø¨Ø· Ø²Ø± Execute Ø¨Ø§Ù„ÙÙ†ÙƒØ´Ù†
    connect(ui->executeBtn, &QPushButton::clicked,
            this, &MainWindow::onExecuteClicked);
}

MainWindow::~MainWindow()
{
    delete ui;
}

// ================================
// LINK BETWEEN GUI <-> DB ENGINE
// ================================
void MainWindow::onExecuteClicked()
{
    // 1ï¸âƒ£ Ù‚Ø±Ø§Ø¡Ø© SQL Ù…Ù† TextEdit
    QString sql = ui->sqlInput->toPlainText().trimmed();
    if (sql.isEmpty())
        return;

    // 2ï¸âƒ£ Reset Ø£ÙŠ Ù†ØªØ§Ø¦Ø¬ Ù‚Ø¯ÙŠÙ…Ø©
    ui->outputBox->clear();
    ui->tableView->setModel(nullptr);

    // 3ï¸âƒ£ ØªÙ†ÙÙŠØ° Ø§Ù„ÙƒÙˆÙŠØ±ÙŠ
    QString result =
        QString::fromStdString(
            db.executeSQLWithResult(sql.toStdString())
            );

    // ðŸ”¥ Ø§Ù„Ø­Ù„ Ø§Ù„Ù…Ù‡Ù… (ÙŠØ­Ø§ÙØ¸ Ø¹Ù„Ù‰ startsWith)
    result = result.trimmed();

    if (result.isEmpty())
        return;

    // ================================
    // CASE 1: RESULT IS A TABLE
    // ================================
    if (result.startsWith("__TABLE__"))
    {
        // Ø¥Ø²Ø§Ù„Ø© Ø§Ù„Ù…Ø§Ø±ÙƒØ±
        result.remove("__TABLE__");
        result = result.trimmed();

        // ØªÙ‚Ø³ÙŠÙ… Ø§Ù„Ø³Ø·ÙˆØ±
        QStringList lines =
            result.split("\n", Qt::SkipEmptyParts);

        if (lines.isEmpty())
            return;

        // Ø¥Ù†Ø´Ø§Ø¡ Model
        QStandardItemModel *model =
            new QStandardItemModel(this);

        // Ø£ÙˆÙ„ Ø³Ø·Ø± = Ø£Ø³Ù…Ø§Ø¡ Ø§Ù„Ø£Ø¹Ù…Ø¯Ø©
        QStringList headers =
            lines.takeFirst().split(" ", Qt::SkipEmptyParts);

        model->setHorizontalHeaderLabels(headers);

        // Ø¨Ø§Ù‚ÙŠ Ø§Ù„Ø³Ø·ÙˆØ± = Ø¨ÙŠØ§Ù†Ø§Øª
        for (int row = 0; row < lines.size(); row++)
        {
            QStringList cols =
                lines[row].split(" ", Qt::SkipEmptyParts);

            for (int col = 0; col < cols.size(); col++)
            {
                model->setItem(
                    row,
                    col,
                    new QStandardItem(cols[col])
                    );
            }
        }

        // Ø±Ø¨Ø· Ø§Ù„Ø¬Ø¯ÙˆÙ„ Ø¨Ø§Ù„Ù€ View
        ui->tableView->setModel(model);

        // ØªØ­Ø³ÙŠÙ† Ø´ÙƒÙ„ Ø§Ù„Ø¬Ø¯ÙˆÙ„
        ui->tableView->horizontalHeader()->setVisible(true);
        ui->tableView->horizontalHeader()
            ->setSectionResizeMode(QHeaderView::Stretch);
    }
    // ================================
    // CASE 2: NORMAL TEXT
    // ================================
    else
    {
        ui->outputBox->append(result);
    }
}
